engine/api/purge/purge_run.go

```diff
@@ -0,0 +1,200 @@
+package purge
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/fsamin/go-dump"
+	"github.com/go-gorp/gorp"
+
+	"github.com/ovh/cds/engine/api/database/gorpmapping"
+	"github.com/ovh/cds/engine/api/repositoriesmanager"
+	"github.com/ovh/cds/engine/api/workflow"
+	"github.com/ovh/cds/engine/cache"
+	"github.com/ovh/cds/engine/featureflipping"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+	"github.com/ovh/cds/sdk/luascript"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/stats"
+)
+
+type MarkAsDeleteOptions struct {
+	DryRun bool
+}
+
+const (
+	RunStatus         = "run_status"
+	RunDateBefore     = "run_days_before"
+	RunGitBranchExist = "git_branch_exist"
+)
+
+func GetRetetionPolicyVariables() []string {
+	return []string{RunDateBefore, RunStatus, RunGitBranchExist}
+}
+
+func markWorkflowRunsToDelete(ctx context.Context, store cache.Store, db *gorp.DbMap, workflowRunsMarkToDelete *stats.Int64Measure) error {
+	dao := new(workflow.WorkflowDAO)
+	wfs, err := dao.LoadAll(ctx, db)
+	if err != nil {
+		return err
+	}
+	for _, wf := range wfs {
+		enabled := featureflipping.IsEnabled(ctx, gorpmapping.Mapper, db, FeaturePurgeName, map[string]string{"project_key": wf.ProjectKey})
+		if !enabled {
+			continue
+		}
+		if _, err := ApplyRetentionPolicyOnWorkflow(ctx, store, db, wf, MarkAsDeleteOptions{DryRun: false}); err != nil {
+			log.ErrorWithFields(ctx, logrus.Fields{"stack_trace": fmt.Sprintf("%+v", err)}, "%s", err)
+		}
+	}
+	workflow.CountWorkflowRunsMarkToDelete(ctx, db, workflowRunsMarkToDelete)
+	return nil
+}
+
+func ApplyRetentionPolicyOnWorkflow(ctx context.Context, store cache.Store, db *gorp.DbMap, wf sdk.Workflow, opts MarkAsDeleteOptions) ([]sdk.WorkflowRunToKeep, error) {
+	runsTokeep := make([]sdk.WorkflowRunToKeep, 0)
+
+	limit := 50
+	offset := 0
+
+	branches, err := getBranchesForWorkflow(ctx, store, db, wf)
+	if err != nil {
+		return nil, err
+	}
+	branchesMap := make(map[string]struct{})
+	for _, b := range branches {
+		branchesMap[b.DisplayID] = struct{}{}
+	}
+
+	for {
+		wfRuns, _, _, count, err := workflow.LoadRuns(db, wf.ProjectKey, wf.Name, offset, limit, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, run := range wfRuns {
+			keep, err := applyRetentionPolicyOnRun(db, wf, run, branchesMap, opts)
+			if err != nil {
+				return nil, err
+			}
+			if keep {
+				runsTokeep = append(runsTokeep, sdk.WorkflowRunToKeep{ID: run.ID, Num: run.Number, Status: run.Status})
+			}
+		}
+
+		if count > offset+limit {
+			offset += limit
+			continue
+		}
+		break
+	}
+	return runsTokeep, nil
+}
+
+func applyRetentionPolicyOnRun(db *gorp.DbMap, wf sdk.Workflow, run sdk.WorkflowRun, branchesMap map[string]struct{}, opts MarkAsDeleteOptions) (bool, error) {
```

Comment:
i don't really understand what the returned boolean means

Reply:
it's the result of retention policy. True means run is kept.